### PR TITLE
TraceSlits class

### PR DESCRIPTION
**Pull request by profxj**
As titled.  This introduces the TraceSlits class
and marks a shift of PYPIT towards 'bite-sized',
container+wrapper objects.  Heaven help us.

Here are the key CHANGES

- Generate arpixels.py module for holding pixel level algorithms
- Move all methods related to TraceSlits to artraceslits.py
- Introduce the TraceSlits class
- Update armlsd accordingly
- Remove driver_trace_slits and refctor_trace_slits methods

@kbwestfall -- Please review the Class
@rcooke-ast  -- Please decide whether arpixels.py is a good idea

### CODE REVIEW

**Comment by rcooke-ast** [comment]
I think it is best to keep all of the code related to the conversion of physical to apparent pixel positions in a separate code, as you have done here. I'm currently of the opinion that we should use apparent pixel positions to reduce the data. Then, once you have a fully reduced 2D science frame, you assign a wavelength distribution to each pixel (which accounts for the spectral tilt and pixel gaps). I now see this as a far off goal... in the meantime, we should probably just stick with apparent pixel positions throughout the entire code to avoid confusion, or sometimes using one and not the other.

**Comment by profxj** [comment]
Ok, @kbwestfall -- I made some 'improvements' to the class
especially for I/O.  Things appear to be passing the dev suite.

I've also added new tests which require downloading 'Cooked'
files into the Dev suite (they should pass otherwise).

I will cease development (unless I see a bug in the Dev tests)
until you review.

File: pypit/armlsd.py [hunk 1]
```diff
@@ -129,39 +133,42 @@ def ARMLSD(fitsdict, reuseMaster=False, reloadMaster=True):
...
+                #                                                      det=det,
+                #                                                      settings=settings.argflag,
+                #                                                      binbpx=slf._bpix[det-1],
+                #                                                      armlsd=True)
+                Tslits = traceslits.TraceSlits(slf._mstrace[det-1], slf._pixlocn[det-1],
+                                               det=det, settings=settings.argflag, binbpx=slf._bpix[det-1])
+                tslits_dict = Tslits.run(armlsd=True)
+
+                # Save in slf
+                slf.SetFrame(slf._lordloc, tslits_dict['lcen'], det)
```
Comment (kbwestfall) [inline pypit/armlsd.py]: I haven't looked at the `TraceSlits` class yet, but is, e.g., `lcen` kept as an attribute of `TraceSlits`?  Maybe you're just trying to keep things separate, but why do we need the `tslits_dict` object?  I think it's cleaner if you have, e.g., `Tslits.lcen` that then gets passed in this call to `SetFrame`.

File: pypit/arpixels.py [hunk 2]
```diff
@@ -0,0 +1,223 @@
+""" Routines related to mapping pixels to physical positions
```
Comment (kbwestfall) [inline pypit/arpixels.py]: Not something we should do now (or possibly ever), but I was planning to suggest that we rethink the general pypit naming convention.  I think X has told me its origin, but I think the inclusion of the 'ar' in the front of every file name is not particularly informative.  I think we should be more intentional about what we call the modules and think through a directory structure that will help organize the modules.  When we/someone imports something from pypit, just the name of the module should give you a good indication of what type of function/object you're importing.  Just a suggestion.

File: pypit/arproc.py [hunk 3]
```diff
@@ -1042,15 +1051,17 @@ def reduce_prepare(slf, sciframe, scidx, fitsdict, det, standard=False):
...
     ###############
     # Subtract off the scattered light from the image
     msgs.work("Scattered light subtraction is not yet implemented...")
     ###############
     # Flat field the science frame (and variance)
     if settings.argflag['reduce']['flatfield']['perform']:
         msgs.info("Flat fielding the science frame")
-        sciframe, rawvarframe = flatfield(slf, sciframe, slf._mspixelflatnrm[det-1], det,
-                                          varframe=rawvarframe, slitprofile=slf._slitprof[det-1])
+        # JXP -- I think it is a bad idea to modify the rawvarframe
```
Comment (kbwestfall) [inline pypit/arproc.py]: Agreed!

File: pypit/scripts/chk_edges.py [hunk 4]
```diff
@@ -33,24 +33,20 @@ def parser(options=None):
...
-    # List only?
-    trc_hdu = fits.open(pargs.file)
-    head0 = trc_hdu[0].header
-
-    mstrace = trc_hdu[0].data
+    # Load up
+    Tslits = traceslits.TraceSlits.from_master_files(pargs.root)
+    lordloc = Tslits.lcen
+    rordloc = Tslits.rcen
+    mstrace = Tslits.mstrace
```
Comment (kbwestfall) [inline pypit/scripts/chk_edges.py]: I'm not sure why this line is yellow.  I might have did something I didn't intend to...

Anyway, I think we should avoid statements like this.  Is it primarily for backwards compatibility?  Within a function, I think we just want to use the `Tslits` object, but we can pass `Tslits.lcen` between functions and call it `lordloc` within that called function.

File: pypit/scripts/show_twod.py [hunk 5]
```diff
@@ -83,12 +84,15 @@ def main(args):
...
-    trc_hdu = fits.open(trc_file)
-    lordloc = trc_hdu[1].data  # Should check name
-    rordloc = trc_hdu[2].data  # Should check name
+    Tslits = traceslits.TraceSlits.from_master_files(trc_file)
+    lordloc = Tslits.lcen
+    rordloc = Tslits.rcen
+
     # Get slit ids
-    stup = (trc_hdu[0].data.shape, lordloc, rordloc)
+    stup = (Tslits.mstrace.shape, lordloc, rordloc)
```
Comment (kbwestfall) [inline pypit/scripts/show_twod.py]: E.g., here I would just call:

```
stup = (Tslits.mstrace.shape, Tslits.lcen, Tslits.rcen)
```

File: pypit/artraceslits.py [hunk 6]
```diff
@@ -0,0 +1,2792 @@
+""" Module for core algorithms related to tracing slits/orders
```
[the remark this answers is not in the document]
Comment (kbwestfall) [reply pypit/artraceslits.py]: This file was mostly just pulling stuff out of artrace.py into a new file, right?  I didn't pay much attention to the details of the code, but I can go back and do this.

File: pypit/traceslits.py [hunk 7]
```diff
@@ -0,0 +1,1023 @@
...
+        # Trace crude me
+        #   -- Mainly to deal with duplicates and improve the traces
+        #   -- Developed for ARMLSD not ARMED
+        if armlsd:
+            self._mslit_tcrude()
+
+        # Synchronize and add in edges
+        if armlsd:
+            self._mslit_sync()
+
```
Comment (kbwestfall) [inline pypit/traceslits.py]: Why are both of these if statements (line 970 and 974) the same?  Is it a bug, or can you just do:

```
if armlsd:
    self._mslit_tcrude()
    self._mslit_sync()
```

File: pypit/armlsd.py (hunk 1 above)
Comment (profxj) [reply pypit/armlsd.py]: I agree with you.  I am forseeing that we will be carrying
around a small object outside of the 'slf' that will contain
the slit info.  And in trying to hold the line about not passing
any non-base object (e.g. dict, array) to any other function
I decided to build this dict.

But, we might well just pass Tslit.

Am going to remove the dict part for now.  @jhennawi will 
be able to handle it.  ;)

Comment (jhennawi) [reply pypit/armlsd.py]: I'm fine with having objects passed into routines provided they are providing value and not adding confusion. I think that the trace slits class is a good example of an object adding value, and indeed we had tset_slits structures in IDL which contained the same information. What I disagree is fancy object oriented stuff where it is not necessary. The arflux module is a good example of this object oriented idea gone wrong. It turns out to be a major struggle to flux the data outside the regular running of PYPIT, which is what a user is inclined to do if they want to change their sensitivity function. But in that case, the driver routine is written with the self object everywhere and relies on being called with PYPIT making the code effectively useless outside of PYPIT.

Comment (profxj) [reply pypit/armlsd.py]: ps.  Am keeping it internally (i.e. the fill_trace_slit_dict method).
Am just not using it anymore.

File: pypit/arpixels.py (hunk 2 above)
Comment (profxj) [reply pypit/arpixels.py]: I agree and am raising it as an issue.

File: pypit/scripts/chk_edges.py (hunk 4 above)
Comment (profxj) [reply pypit/scripts/chk_edges.py]: expunged

File: pypit/scripts/show_twod.py (hunk 5 above)
Comment (profxj) [reply pypit/scripts/show_twod.py]: agreed.  modified.

File: pypit/artraceslits.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/artraceslits.py]: And so began core/
which I think is a nice way
to get us to think more carefully about
what is core and what isn't.

I am still calling it artraceslits.py
but that can change (someday).

File: pypit/traceslits.py [hunk 8]
```diff
@@ -0,0 +1,1023 @@
...
+trace_slits_dict = {}
+trace_slits_dict['lcen'] = None
+trace_slits_dict['rcen'] = None
+trace_slits_dict['lordpix'] = None
+trace_slits_dict['rordpix'] = None
+trace_slits_dict['slitpix'] = None
+trace_slits_dict['pixcen'] = None
+trace_slits_dict['pixwid'] = None
+trace_slits_dict['extrapord'] = None
+
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/traceslits.py]: It is/was never used.  
Removed.

File: pypit/traceslits.py (hunk 7 above)
Comment (profxj) [reply pypit/traceslits.py]: Combined.
Comments too.

File: pypit/traceslits.py [hunk 9]
```diff
@@ -0,0 +1,1023 @@
...
+        Returns
+        -------
+        self.trace_slits_dict
+
+        """
+        self.trace_slits_dict = {}
+        for key in ['lcen', 'rcen', 'pixcen', 'pixwid', 'lordpix',
+                    'rordpix', 'extrapord', 'slitpix']:
+            self.trace_slits_dict[key] = getattr(self, key)
+        return self.trace_slits_dict
```
[the remark this answers is not in the document]
Comment (profxj) [reply pypit/traceslits.py]: Sorry, that was confusing. 
Am keeping this method but it is not used
for the time being.

**Comment by profxj** [comment]
passing all Dev tests.

Merging

**Pull request merged**